README


                _     .---, .---,------.------.
               | '-.  |   | |   |      |  _   |
	.------'    '-.   | |   |  ---<  '-'  |
	|              >  |__   |      |   .--'
	'------,    ,-'      \  |>---  |   |
	       |_,-'  |______/__:______'___:
 

	ArrowLISP - An interpreter for Purely Symbolic LISP
	Copyright (C) 1998-2006 Nils M Holm. All rights reserved.
	See the file LICENSE for conditions of use.


	--== Latest News ==--

	The most notable change in this version is the introduction
	of the VERIFY-ARROWS meta function. Arrow verification makes
	sure that N is the normal form of X in X => N. See the
	manual for details.

	Logical falsity is now expressed using :F (truth is T).
	Sorry about this drastic change, but having an extra
	symbol for falsity simplifies the implementation of
	some algorithms significantly.

	Added ArrowLISP Micro KANREN, an implementation of
	KANREN-like logic programming extensions. See src/amk/
	for source code and examples.


	--== What is ArrowLISP? ==--

	ArrowLISP is an interpreter for a purely symbolic,
	side effect-free, lexically scoped dialect of LISP.
	It may be considered an implementation of pure LISP
	plus global definitions.


	--== Installation ==--

	The interpreter is written in ANSI C (C89) and LISP.
	It should compile without any modifications on a
	variety of systems including *BSD, Linux, Plan 9
	(using APE), and W*ndows via MinGW or Cygwin.

	On a typical post-1990 Unix system typing

	make
	su
	make install
	exit

	should work fine.


	--== Getting Started ==--

	After installing ArrowLISP, type

	alisp

	to launch the interpreter.

	Typing

	man 7 alisp

	will produce the alisp(7) man page.

	The Reference Manual can be found in the file

	/usr/local/share/doc/alisp/alisp.txt

	(unless you are using a different PREFIX).


	--== Feedback ==--

	Please send your input to

	Nils M Holm <nmh@t3x.org>

	Thanks!



Changes
2006-09-17 Fixed false positive in test suite.
2006-09-13 Fixed tail recursion while passing closures.
2006-09-12 Added GENERATOR example (src/misc/gener.l)
2006-09-12 Updated reference manual (alisp.txt) and man page (alisp.7).
2006-09-11 Fixed natural = operator (did not normalize arguments).
2006-09-09 Made {...} unREADable.
2006-09-09 Changed representation of unreadable objects to {...}.
2006-09-09 Fixed xread(): no longer returns EOT on bad pairs.
2006-09-08 Fixed ()=>() whith VERIFY-ARROWS enabled.
2006-09-07 Added -L option (print license).
2006-09-07 Renamed examples/ to src/.
2006-09-07 ALISP_RELEASE now reflects all changes (not just alisp.c).
2006-09-04 Added a more general predicate iterator (iter.l).
2006-08-27 Updated LVRENAME and UNLET programs (src/misc).
2006-08-22 Made NUMBER-P catch dotted lists.
2006-08-22 Made INTEGER reduce rationals when RMATH is loaded.
2006-08-17 Added REDUCE-R function (right-associative REDUCE).
2006-08-15 Closures no longer capture quoted symbols (see collect()).
2006-07-16 Added ArrowLISP Micro KANREN and ZEBRA example.
2006-07-13 Made ALISP restore default package when aborting evaluation.
2006-07-01 Fixed CopyBindings() (did not copy all packages).
2006-06-30 Changed representation of false value from () to :F.
2006-06-30 Made :T an alias of T.
2006-06-27 Allowed (EXPLODE ()) => () and (IMPLODE ()) => ().
2006-06-26 Made WRITE print a leading quote.
2006-06-16 Added QUIT primitive.
2006-06-15 Improved some error messages.
2006-06-14 Added -g (track GC) command line option.
2006-06-08 Made application of CAR and CDR on internals illegal.
2006-06-08 GC-Protected Error Context (only in case of an error).
2006-06-08 GC-Protected expressions during EVAL. Oops.


shell.c
/*
 * ArrowLISP -- An interpreter for purely symbolic LISP
 * Copyright (C) 2005,2006 Nils M Holm <nmh@t3x.org>
 * This is the interpreter shell.
 * See the file LICENSE for conditions of use.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include "alisp.h"

#define STRLEN	256

char	Image[STRLEN];
int	Nodes;
int	Batch;
int	LoudGC;

void	catchIntr(int sig);
int	getOptNum(int argc, char **argv, int *pi, int *pj, int *pk);
void	getOpts(int argc, char **argv);
void	help(void);
void	init(void);
void	license(void);
void	repl(void);
void	usage(void);

/* Print usage and fail */
void usage(void) {
	fprintf(stderr,
		"Usage: alisp [-L] [-bgi] [-n nodes] [image]\n");
}

int getOptNum(int argc, char **argv, int *pi, int *pj, int *pk) {
	int	n, c;

	if (++(*pi) >= argc) {
		usage();
		exit(1);
	}
	n = atoi(argv[*pi]);
	c = argv[*pi][strlen(argv[*pi])-1];
	switch (c) {
	case 'K':	n = n * 1024; break;
	case 'M':	n = n * 1024 * 1024; break;
	}
	*pj = *pk = 0;
	return n;
}

void help(void) {
	fputc('\n', stderr);
	usage();
	fprintf(stderr,
		"\n"
		"-b    batch mode (quiet, exit on first error)\n"
		"-g    report number of free nodes after each GC\n"
		"-i    init mode (do not load any image)\n"
		"-n #  number of nodes to allocate (default: %dK)\n"
		"-L    print license and exit\n"
		"\n"
		"default image: %s\n\n",
		ALISP_DFL_NODES/1024, ALISP_DFL_IMAGE);
}

void license(void) {
	char	**s;

	s = alisp_license();
	while (*s) {
		printf("%s\n", *s);
		s++;
	}
	exit(0);
}

/* Evaluate command line options */
void getOpts(int argc, char **argv) {
	char	*a;
	int	i, j, k;
	int	v;

	strncpy(Image, ALISP_DFL_IMAGE, strlen(ALISP_DFL_IMAGE));
	Image[STRLEN-1] = 0;
	Nodes = ALISP_DFL_NODES;
	LoudGC = 0;
	v = 0;
	i = 1;
	while (i < argc) {
		a = argv[i];
		if (a[0] != '-') break;
		k = strlen(a);
		for (j=1; j<k; j++) {
			switch (a[j]) {
			case 'b':
				Batch = 1;
				break;
			case 'n':
				Nodes = getOptNum(argc, argv, &i, &j, &k);
				break;
			case 'g':
				LoudGC = 1;
				break;
			case 'i':
				Image[0] = 0;
				break;
			case 'L':
				license();
				break;
			case '?':
			case 'h':
				help();
				exit(1);
				break;
			default:
				usage();
				exit(1);
			}
		}
		i = i+1;
	}
	if (i < argc) {
		strncpy(Image, a, strlen(a)+1);
		Image[STRLEN-1] = 0;
	}
	if (Nodes < ALISP_MIN_SIZE) {
		fprintf(stderr, "alisp: minimum pool size is %d\n",
			ALISP_MIN_SIZE);
		exit(1);
	}
}

void catchIntr(int sig) {
	sig = 0; /*LINT*/
	alisp_stop();
	signal(SIGINT, catchIntr);
}

void repl(void) {
	int	n;

	while(1) {
		alisp_reset();
		n = alisp_read();
		if (alisp_eof(n)) return;
		if (alisp_errflag()) {
			printf("* ");
			alisp_print_error();
			if (Batch) exit(1);
			continue;
		}
		n = alisp_eval(n);
		if (alisp_errflag()) {
			printf("* ");
			alisp_print_error();
			if (Batch) exit(1);
		}
		else {
			if (!Batch) printf("=> ");
			alisp_print(n);
			alisp_nl();
		}
	}
}

void init(void) {
	if (alisp_init(Nodes, LoudGC)) {
		fprintf(stderr, "alisp init failed (memory problem)\n");
		exit(1);
	}
}

/*
 * Here the fun begins
 */
int main(int argc, char **argv) {
	getOpts(argc, argv);
	init();
	getOpts(argc, argv);
	if (!Batch) {
		printf("ArrowLISP ");
		printf(ALISP_RELEASE);
		printf(" (C) 2006 Nils M Holm\n");
	}
	if (Image[0]) {
		if (alisp_load_image(Image)) {
			printf("* ");
			alisp_print_error();
			if (Batch) exit(1);
			alisp_fini();
			init();
			getOpts(argc, argv);
		}
	}
	else if (!Batch) {
		printf("Warning: no image loaded\n");
	}
	signal(SIGINT, catchIntr);
	repl();
	alisp_fini();
	return 0;
}


alisp.h
/*
 * ArrowLISP -- An interpreter for purely symbolic LISP
 * Copyright (C) 2005,2006 Nils M Holm <nmh@t3x.org>
 * See http://t3x.org/alisp/license.html for conditions of use.
 */

#define ALISP_MAJOR	17
#define ALISP_RELEASE	"2006-09-13"

/*
 * Number of nodes and vector cells.
 * Memory = Nodes * (2 * sizeof(int) + 1) 
 * (Recommended: 16-bit: 12280, >=32-bit: 131072)
 */
#define	ALISP_DFL_NODES	131072

/* The node pool should not be smaller than this. */
#define ALISP_MIN_SIZE	12280

/* Default image */
#define ALISP_DFL_IMAGE	"/usr/local/share/alisp/alisp-image"

/* For errors without specific source. */
#define ALISP_NOEXPR	-1

/* Counter for statistics. */
struct counter {
	int	n,	/* ones */
		n1k,	/* thousands */
		n1m,	/* millions */
		n1g;	/* billions */
};

/* GC statistics */
struct gc_stats {
	int	used,	/* nodes currently used */
		peak;	/* peak use */
};

/* Error reporting */
struct errorContext {
	char	*msg;	/* Most recent error message */
	char	*arg;	/* Additional information */
	int	expr;	/* Expression causing last error */
	char	*file;	/* File of last error */
	int	line;	/* Line number of last error */
	int	fun;	/* Function of last error */
	int	frame;	/* Call frame of last error */
};

#define alisp_eof(n)	((n) < 0)

int	alisp_dump_image(char *p);
struct errorContext
	*alisp_errctx(void);
int	alisp_errflag(void);
int	alisp_eval(int n);
void	alisp_fini(void);
int	alisp_init(int nodes, int trackGc);
char	**alisp_license(void);
int	alisp_load_image(char *p);
void	alisp_nl(void);
void	alisp_print(int n);
void	alisp_print_error(void);
int	alisp_read(void);
void	alisp_reset(void);
void	alisp_stop(void);
